OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -433,33 +443,33 @@ protected void CancelProduction(string itemName, uint numberToCancel)
 
 		bool CancelProductionInner(string itemName)
 		{
-			var lastIndex = queue.FindLastIndex(a => a.Item == itemName);
+			var item = Queue.LastOrDefault(a => a.Item == itemName);
 
-			if (lastIndex > 0)
-				queue.RemoveAt(lastIndex);
-			else if (lastIndex == 0)
+			if (item != null)
 			{
-				var item = queue[0];
-
 				// Refund what has been paid
 				playerResources.GiveCash(item.TotalCost - item.RemainingCost);
-				FinishProduction();
+				FinishProduction(item);
 			}
 			else
 				return false;
 
 			return true;
 		}
 
-		public void FinishProduction()
+		public void FinishProduction(ProductionItem item)
```

Comment:
This name is no longer accurate - you call this on cancel as well as finish.  How about `EndProduction`?